pgjdbc/src/test/java/org/postgresql/test/jdbc2/TimezoneCachingTest.java

```diff
@@ -0,0 +1,341 @@
+   */
+  public void testResultSetCachedTimezoneInstance() throws SQLException {
+    Timestamp ts = new Timestamp(2016 - 1900, 1 - 1, 31, 0, 0, 0, 0);
+    Statement stmt = null;
+    PreparedStatement pstmt = null;
+    ResultSet rs = null;
+    try {
+      pstmt = con.prepareStatement("INSERT INTO testtz VALUES (?,?)");
+      pstmt.setInt(1, 1);
+      pstmt.setTimestamp(2, ts);
+      pstmt.addBatch();
+      pstmt.executeBatch();
+      stmt = con.createStatement();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getInt(1);
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getTimestamp(2);
+      assertTrue("Cache should be set", getTimeZoneCache(rs) != null);
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getInt(1);
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getObject(2);
+      assertTrue("Cache should be set", getTimeZoneCache(rs) != null);
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getInt(1);
+      assertTrue("Cache should NOT be set", getTimeZoneCache(rs) == null);
+      rs.getDate(2);
+      assertTrue("Cache should be set", getTimeZoneCache(rs) != null);
+      rs.close();
+    } finally {
+      TestUtil.closeQuietly(rs);
+      TestUtil.closeQuietly(pstmt);
+      TestUtil.closeQuietly(stmt);
+    }
+  }
+
+  /**
+   * Test to check the internal cached timezone of a result set is used as expected.
+   */
+  public void testResultSetCachedTimezoneUsage() throws SQLException {
+    Timestamp ts = new Timestamp(2016 - 1900, 1 - 1, 31, 0, 0, 0, 0);
+    Statement stmt = null;
+    PreparedStatement pstmt = null;
+    ResultSet rs = null;
+    TimeZone tz1 = TimeZone.getTimeZone("GMT+8:00");
+    TimeZone tz2 = TimeZone.getTimeZone("GMT-2:00");
+    Calendar c1 = new GregorianCalendar(tz1);
+    try {
+      TimeZone.setDefault(tz1);
+      pstmt = con.prepareStatement("INSERT INTO testtz VALUES (?,?)");
+      pstmt.setInt(1, 1);
+      pstmt.setTimestamp(2, ts);
+      pstmt.addBatch();
+      pstmt.executeBatch();
+      stmt = con.createStatement();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      assertTrue("Timestamps sould be in same time zone", rs.getTimestamp(2).equals(ts));
+      rs.close();
+      rs = stmt.executeQuery("SELECT col1, col2 FROM testtz");
+      rs.next();
+      rs.getInt(1);
+      TimeZone.setDefault(tz2);
+      assertTrue("Timestamps sould NOT be in same time zone", !rs.getTimestamp(2).equals(ts));
```

Comment:
This is "happy debugging" test anti-pattern.

Please convert `assertTrue` to `assertEquals` as described here: https://github.com/pgjdbc/pgjdbc#test


"Timestamps should NOT be in same time zone" vs "timestamp was passed as with TZ1 (default at time of execute) and default zone was changed to TZ2, thus the resulting value should be ...."
Well, it takes time to produce those messages, however they are extremely valuable.
E.g. https://github.com/pgjdbc/pgjdbc/pull/580/files#diff-efc8c3eee49c38f1557353bff58b7e94R1238


I mean the following: either use "why" kind of messages, or remove the messages and just use `assertEquals`. Of course, whys are better.